pkg/adapter/apiserver/adapter.go

```diff
@@ -17,113 +17,160 @@ limitations under the License.
-func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
-	// Convert the namespace/name string into a distinct namespace and name
-	namespace, name, err := cache.SplitMetaNamespaceKey(key)
-	if err != nil {
-		r.Logger.Errorf("invalid resource key: %s", key)
-		return nil
+func NewAdaptor(source string, k8sClient dynamic.Interface, ceClient cloudevents.Client, logger *zap.SugaredLogger, opt Options) Adapter {
+	mode := opt.Mode
+	switch mode {
+	case ResourceMode, RefMode:
+		// ok
+	default:
+		logger.Warn("unknown mode ", mode)
+		mode = RefMode
+		logger.Warn("defaulting mode to ", mode)
 	}
 
-	// Get the resource with this namespace/name
-	original, err := r.lister.ByNamespace(namespace).Get(name)
-	if apierrs.IsNotFound(err) {
-		// The resource may no longer exist, in which case we stop processing.
-		r.Logger.Error("resource key in work queue no longer exists", zap.Any("key", key))
-		return nil
-	} else if err != nil {
-		return err
+	a := &adapter{
+		k8s:       k8sClient,
+		ce:        ceClient,
+		source:    source,
+		logger:    logger,
+		gvrcs:     opt.GVRCs,
+		namespace: opt.Namespace,
+		mode:      mode,
 	}
+	return a
+}
 
-	object := original.(*duckv1alpha1.AddressableType)
+type eventDelegate interface {
+	addEvent(obj interface{})
+	updateEvent(oldObj, newObj interface{})
+	deleteEvent(obj interface{})
 
-	eventType := updateEventType
-	timestamp := object.GetCreationTimestamp()
-	if object.GetDeletionTimestamp() != nil {
-		eventType = deleteEventType
-		timestamp = *object.GetDeletionTimestamp()
-	}
+	addControllerWatch(gvr schema.GroupVersionResource)
+}
 
-	objectRef := corev1.ObjectReference{
-		APIVersion: object.APIVersion,
-		Kind:       object.Kind,
-		Name:       object.GetName(),
-		Namespace:  object.GetNamespace(),
+func (a *adapter) Start(stopCh <-chan struct{}) error {
+	// Local stop channel.
+	stop := make(chan struct{})
+
+	resyncPeriod := time.Duration(10 * time.Hour)
+
+	var d eventDelegate
+	switch a.mode {
+	case ResourceMode:
+		d = &resource{
+			ce:     a.ce,
+			source: a.source,
+			logger: a.logger,
+		}
+
+	case RefMode:
+		d = &ref{
+			ce:     a.ce,
+			source: a.source,
+			logger: a.logger,
+		}
+
+	default:
+		return fmt.Errorf("mode %q not understood", a.mode)
 	}
 
-	event := cloudevents.Event{
-		Context: cloudevents.EventContextV02{
-			ID:     string(object.GetUID()),
-			Type:   eventType,
-			Source: *types.ParseURLRef(object.GetSelfLink()),
-			Time:   &types.Timestamp{Time: timestamp.Time},
-		}.AsV02(),
-		Data: objectRef,
+	for _, gvrc := range a.gvrcs {
+		var informer cache.SharedIndexInformer
+
+		lw := &cache.ListWatch{
+			ListFunc:  asUnstructuredLister(a.k8s.Resource(gvrc.GVR).Namespace(a.namespace).List),
+			WatchFunc: asUnstructuredWatcher(a.k8s.Resource(gvrc.GVR).Namespace(a.namespace).Watch),
+		}
+		informer = cache.NewSharedIndexInformer(lw, &unstructured.Unstructured{}, resyncPeriod, cache.Indexers{
```

Comment:
@n3wscott  have you considered getting rid of the informer altogether and use only `Reflector`?

Reply:
This is a great idea. Thanks! Worked it in and updated.

Reply:
My understanding is that the informer will remove some possible duplicate information and give some serialization for us, but mayhaps for what we're doing here it's fine.